src/wovnio/wovnphp/Headers.php

```diff
@@ -315,4 +328,29 @@ public function getDocumentURI()
 
         return $this->removeLang($url, $this->requestLang());
     }
+
+    public function constructOriginalURL()
+    {
+        $originalURL = $this->protocol . '://' . $this->originalHost . $this->originalPath;
+        if ($this->query) {
+            $originalURL = $originalURL . '?' . $this->query;
+        }
+        return $originalURL;
+    }
+
+    public function computeRedirectUrl()
+    {
+        $cookieLangCode = $this->cookieLang->getCookieLang();
+        $url = $this->urlKeepTrailingSlash;
+        $url = Url::addLangCode($url, $this->store, $cookieLangCode, $this);
+        return $url;
+    }
+
+    public function shouldRedirect()
+    {
+        if (!$this->store->settings['use_cookie_lang']) {
+            return false;
+        }
+        return $this->cookieLang->getCookieLang() && ($this->requestLang() != $this->cookieLang->getCookieLang()) && $this->requestLang() === $this->store->defaultLang();
```

Comment:
extract `$cookieLang, $requestLang` variables instead of calling these twice. I think doing the requestLang is default check first might make more sense.
```

return $requestLang === $this->store->defaultLang()
	&& $cookieLang
	&& $this->requestLang() != $cookieLang
```

Reply:
actually I guess the $cookieLang check is the simplest so how you had it is fine. Take this as a formatting suggestion though.